### ISSUE TITLE
How to customize PATH for sensu client?

### ISSUE THREAD

**Issue by WayneYe** [sensu/sensu#1064]
I have a sensu client, I debugged its PATH in one of my check script, and it shows:

```
/sbin:/usr/sbin:/bin:/usr/bin:/etc/sensu/plugins:/etc/sensu/handlers
```

How can I customize this PATH for sensu, say: I want to add /usr/local/bin to the end of the PATH, which results in: 

```
/sbin:/usr/sbin:/bin:/usr/bin:/etc/sensu/plugins:/etc/sensu/handlers:/usr/local/bin
```

I've tried many ways but didn't success, I've tried:
1. Set sensu user's SHELL to /bin/bash (instead of the default /bin/false), and add .bashrc|.profile under sensu user's home dir /opt/sensu, with this line: export PATH=$PATH:/usr/local/bin
2. Edit /etc/default/sensu, add this line PATH=$PATH:/usr/local/bin
3. By reading this: https://sensuapp.org/docs/latest/clients, I set USER=ec2-user in /etc/default/sensu, after restarting sensu client, I clearly see the sensu client process is running by ec2-user, however, surprisingly, the PATH is NOT the same as ec2-user

All 1,2 and 3 above didn't work, in my check script written in python, I have those lins:

``` python
from subprocess import call, Popen, PIPE
import os
import sys
import shlex

import platform
print os.environ["PATH"]

proc = Popen(['which', 'python'],
        stdout=PIPE,
        stderr=PIPE)
out, err = proc.communicate() #does not return until the process has terminated.
print(out)
print(err)
#print(platform.__dict__)
print(platform.python_version())

proc = Popen(['whoami'],
        stdout=PIPE,
        stderr=PIPE)
out, err = proc.communicate() #does not return until the process has terminated.
print(out)
print(err)
sys.exit(0)
```

The output was:

``` shell
/sbin:/usr/sbin:/bin:/usr/bin:/etc/sensu/plugins:/etc/sensu/handlers
/usr/bin/python   
2.6.6 
ec2-user
```

Update, while I wrote this line in my python check script:

``` python
proc = Popen(['bash','--login', '-x'], stdout=PIPE, stderr=PIPE)
out, err = proc.communicate()
print(out)
print(err)
```

I saw it output:

```
PATH=/sbin:/usr/sbin:/bin:/usr/bin:/etc/sensu/plugins:/etc/sensu/handlers:/usr/local/sbin:/usr/local/bin
```

However, the other output surprisingly.... "which python" -> /usr/bin/python, "python --version" -> 2.6.6
Request help....

**Comment by portertech** [comment]
@WayneYe have you tried exporting PATH in `/etc/default/sensu`? For example: `export PATH=$PATH:/usr/local/bin`

**Comment by WayneYe** [comment]
@portertech what you commented is exactly my No.2 attempt above...

I really don't understand why this simply task can't be accomplished, running py scripts is so common.

**Comment by portertech** [comment]
@WayneYe `PATH=$PATH:/usr/local/bin` is not the same as `export PATH=$PATH:/usr/local/bin`. Did you export PATH?

**Comment by portertech** [comment]
I suspect that this issue may have to do w/ the source'n of Sensu service config files, https://github.com/sensu/sensu-build/blob/master/sensu_configs/init.d/sensu-service#L30 vs https://github.com/sensu/sensu-build/blob/master/sensu_configs/init.d/sensu-service#L104

**Comment by dmsimard** [comment]
So I tried to add PATH=$PATH:/usr/local/bin to /etc/default/sensu but it doesn't work.

From looking at the output of bash -x /etc/init.d/sensu-service server restart, we can see that the PATH variable is being set as expected from sourcing /etc/default/sensu but it is later overwritten by /etc/init.d/functions:
```
# bash -x /etc/init.d/sensu-service server restart
+ sensu_service=sensu-server
+ EMBEDDED_RUBY=true
+ CONFIG_FILE=/etc/sensu/config.json
+ CONFIG_DIR=/etc/sensu/conf.d
+ EXTENSION_DIR=/etc/sensu/extensions
+ PLUGINS_DIR=/etc/sensu/plugins
+ HANDLERS_DIR=/etc/sensu/handlers
+ LOG_DIR=/var/log/sensu
+ LOG_LEVEL=info
+ PID_DIR=/var/run/sensu
+ USER=sensu
+ SERVICE_MAX_WAIT=10
+ '[' -f /etc/default/sensu ']'
+ . /etc/default/sensu                                              # <-- Sourcing /etc/default/sensu
++ EMBEDDED_RUBY=false
++ LOG_LEVEL=info
++ SERVICE_MAX_WAIT=10
++ PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/bin:/usr/local/bin # <-- PATH set correctly
+ system=unknown
+ '[' -f /etc/redhat-release ']'
+ system=redhat
+ '[' redhat = redhat ']'
+ . /etc/init.d/functions                                           # <-- Sourcing /etc/init.d/functions
++ TEXTDOMAIN=initscripts
++ umask 022
++ PATH=/sbin:/usr/sbin:/bin:/usr/bin                               # <-- Overrides whatever is set and re-exports
++ export PATH
++ '[' 16323 -ne 1 -a -z '' ']'
++ /bin/mountpoint -q /cgroup/systemd
++ /bin/mountpoint -q /sys/fs/cgroup/systemd
++ case "$0" in
++ _use_systemctl=1
++ '[' -z '' ']'
++ COLUMNS=80
++ '[' -z '' ']'
++ '[' -c /dev/stderr -a -r /dev/stderr ']'
+++ /sbin/consoletype
++ CONSOLETYPE=pty
++ '[' -z '' ']'
++ '[' -z '' ']'
++ '[' -f /etc/sysconfig/i18n -o -f /etc/locale.conf ']'
++ . /etc/profile.d/lang.sh
++ unset LANGSH_SOURCED
++ '[' -z '' ']'
++ '[' -f /etc/sysconfig/init ']'
++ . /etc/sysconfig/init
+++ BOOTUP=color
+++ RES_COL=60
+++ MOVE_TO_COL='echo -en \033[60G'
+++ SETCOLOR_SUCCESS='echo -en \033[0;32m'
+++ SETCOLOR_FAILURE='echo -en \033[0;31m'
+++ SETCOLOR_WARNING='echo -en \033[0;33m'
+++ SETCOLOR_NORMAL='echo -en \033[0;39m'
++ '[' pty = serial ']'
++ __sed_discard_ignored_files='/\(~\|\.bak\|\.orig\|\.rpmnew\|\.rpmorig\|\.rpmsave\)$/d'
++ '[' 1 = 1 ']'
++ '[' xserver = xstart -o xserver = xstop -o xserver = xrestart -o xserver = xreload -o xserver = xtry-restart -o xserver = xforce-reload -o xserver = xcondrestart ']'
+ '[' -r /etc/sysconfig/sensu-server ']'
+ lockfile=/var/lock/subsys/sensu-server
+ '[' redhat = debian ']'
+ '[' redhat = suse ']'
+ cd /opt/sensu
+ exec=/opt/sensu/bin/sensu-server
+ pidfile=/var/run/sensu/sensu-server.pid
+ logfile=/var/log/sensu/sensu-server.log
+ options='-b -c /etc/sensu/config.json -d /etc/sensu/conf.d -e /etc/sensu/extensions -p /var/run/sensu/sensu-server.pid -l /var/log/sensu/sensu-server.log -L info '
+ case "$2" in
+ restart
+ stop
+ '[' x = xtrue ']'
+ log_daemon_msg 'Stopping sensu-server'
+ echo -n 'Stopping sensu-server'
Stopping sensu-server++ pid_pgrep
++ pgrep -f -P 1 -u sensu ' /opt/sensu/bin/sensu-server '
+ pid=26527
+ '[' '!' -n 26527 ']'
+ '[' -n 26527 ']'
+ kill 26527
+ retval=0
+ '[' 0 -eq 0 ']'
+ wait_for_stop
+ stopped=1
+ attempts=0
+ '[' 0 -lt 10 ']'
+ attempts=1
+ status
+ '[' 0 -eq 0 ']'
+ sleep 1
+ '[' 1 -lt 10 ']'
+ attempts=2
+ status
+ '[' 0 -eq 0 ']'
+ sleep 1
+ '[' 2 -lt 10 ']'
+ attempts=3
+ status
+ '[' 1 -eq 0 ']'
+ rm -f /var/run/sensu/sensu-server.pid
+ rm -f /var/lock/subsys/sensu-server
+ stopped=0
+ break
+ return 0
+ retval=0
+ '[' 0 -eq 0 ']'
+ echo_ok
+ echo_success
+ '[' color = color ']'
+ echo -en '\033[60G'
                                                           + echo -n '['
[+ '[' color = color ']'
+ echo -en '\033[0;32m'
+ echo -n '  OK  '
  OK  + '[' color = color ']'
+ echo -en '\033[0;39m'
+ echo -n ']'
]+ echo -ne '\r'
+ return 0
+ echo

+ return 0
+ retval=0
+ '[' 0 -eq 0 ']'
+ start
+ log_daemon_msg 'Starting sensu-server'
+ echo -n 'Starting sensu-server'
Starting sensu-server+ '[' -x /opt/sensu/bin/sensu-server ']'
+ status
+ '[' 3 -eq 0 ']'
+ set_sensu_paths
+ '[' xfalse = xtrue ']'
+ export PATH=/sbin:/usr/sbin:/bin:/usr/bin:/etc/sensu/plugins:/etc/sensu/handlers
+ PATH=/sbin:/usr/sbin:/bin:/usr/bin:/etc/sensu/plugins:/etc/sensu/handlers
+ ensure_dir /var/run/sensu sensu
+ '[' '!' -d /var/run/sensu ']'
+ ensure_dir /var/log/sensu sensu
+ '[' '!' -d /var/log/sensu ']'
+ start_daemon sensu /var/run/sensu/sensu-server.pid /opt/sensu/bin/sensu-server '-b -c /etc/sensu/config.json -d /etc/sensu/conf.d -e /etc/sensu/extensions -p /var/run/sensu/sensu-server.pid -l /var/log/sensu/sensu-server.log -L info '
+ daemon --user sensu --pidfile /var/run/sensu/sensu-server.pid '/opt/sensu/bin/sensu-server -b -c /etc/sensu/config.json -d /etc/sensu/conf.d -e /etc/sensu/extensions -p /var/run/sensu/sensu-server.pid -l /var/log/sensu/sensu-server.log -L info '
+ local gotbase= force= nicelevel corelimit
+ local pid base= user= nice= bg= pid_file=
+ local cgroup=
+ nicelevel=0
+ '[' --user '!=' -user ']'
+ case $1 in
+ user=sensu
+ shift 2
+ '[' --pidfile '!=' -pidfile ']'
+ case $1 in
+ pid_file=/var/run/sensu/sensu-server.pid
+ shift 2
+ '[' '/opt/sensu/bin/sensu-server -b -c /etc/sensu/config.json -d /etc/sensu/conf.d -e /etc/sensu/extensions -p /var/run/sensu/sensu-server.pid -l /var/log/sensu/sensu-server.log -L info ' '!=' '/opt/sensu/bin/sensu-server -b -c /etc/sensu/config.json -d /etc/sensu/conf.d -e /etc/sensu/extensions -p /var/run/sensu/sensu-server.pid -l /var/log/sensu/sensu-server.log -L info ' ']'
+ '[' -z '' ']'
+ base='sensu-server.log -L info '
+ __pids_var_run 'sensu-server.log -L info ' /var/run/sensu/sensu-server.pid
+ local 'base=sensu-server.log -L info '
+ local pid_file=/var/run/sensu/sensu-server.pid
+ pid=
+ '[' -f /var/run/sensu/sensu-server.pid ']'
+ return 3
+ '[' -n '' -a -z '' ']'
+ corelimit='ulimit -S -c 0'
+ '[' -n '' ']'
+ '[' -n '' ']'
+ '[' color = verbose -a -z '' ']'
+ '[' -z sensu ']'
+ runuser -s /bin/bash sensu -c 'ulimit -S -c 0 >/dev/null 2>&1 ; /opt/sensu/bin/sensu-server -b -c /etc/sensu/config.json -d /etc/sensu/conf.d -e /etc/sensu/extensions -p /var/run/sensu/sensu-server.pid -l /var/log/sensu/sensu-server.log -L info '
+ '[' 0 -eq 0 ']'
+ success 'sensu-server.log -L info  startup'
+ '[' color '!=' verbose -a -z '' ']'
+ echo_success
+ '[' color = color ']'
+ echo -en '\033[60G'
                                                           + echo -n '['
[+ '[' color = color ']'
+ echo -en '\033[0;32m'
+ echo -n '  OK  '
  OK  + '[' color = color ']'
+ echo -en '\033[0;39m'
+ echo -n ']'
]+ echo -ne '\r'
+ return 0
+ return 0
+ retval=0
+ sleep 1
+ status
+ '[' 0 -ne 0 ']'
+ '[' 0 -eq 0 ']'
+ touch /var/lock/subsys/sensu-server
+ echo_ok
+ echo_success
+ '[' color = color ']'
+ echo -en '\033[60G'
                                                           + echo -n '['
[+ '[' color = color ']'
+ echo -en '\033[0;32m'
+ echo -n '  OK  '
  OK  + '[' color = color ']'
+ echo -en '\033[0;39m'
+ echo -n ']'
]+ echo -ne '\r'
+ return 0
+ echo

+ return 0
+ exit 0
```

This is what /etc/init.d/functions looks like: https://gist.github.com/dmsimard/fc9f8019dbdf33f8f159 (CentOS 7.2)

**Comment by calebhailey** [comment]
Tagging this for consideration in the 0.24 release.

**Comment by calebhailey** [comment]
The 0.25 release is going to hit today with fewer features than originally anticipated & primarily only some internal improvements. Moving this issue to 0.26.

**Comment by portertech** [comment]
@amdprophet @cwjohnston I suspect that this is due to the service script controlling $PATH for the sensu processes. This may change with the significant changes to packaging and service management for 0.27.

**Comment by amdprophet** [comment]
@WayneYe @dmsimard can you confirm if this issue is still occurring for you with Sensu 0.28.4?

I ran a few tests and it looks like PATH can be modified in `/etc/default/sensu` without exporting it. The binstub located in `/opt/sensu/bin/sensu-service` will append the Sensu plugin & handler directories to the end of PATH but you should still be able to modify most of the PATH.

**Comment by portertech** [comment]
Fixed in Sensu >= 0.27
